File: Kruskal算法求最小生成树.cpp
#include <bits/stdc++.h>
using namespace std;

#define ff first
#define ss second
#define int long long

struct DSU { // 并查集
    vector<int> pre, siz;
    
    DSU() {}
    DSU(int n) {
        pre.resize(n);
        iota(pre.begin(), pre.end(), 0);
        siz.assign(n, 1);
    }
    
    int find(int x) {
        if (pre[x] == x) {
            return x;
        }
        return pre[x] = find(pre[x]);
    }
    
    bool same(int x, int y) {
        return find(x) == find(y);
    }
    
    bool merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return false;
        }
        siz[x] += siz[y];
        pre[y] = x;
        return true;
    }
    
    int size(int x) {
        return siz[find(x)];
    }
};

signed main() {
    int n, m, ans = 0, cnt = 0;
    cin >> n >> m;
    DSU dsu(n + 1);
    vector<pair<int, pair<int, int>>> edges(m);
    for (pair<int, pair<int, int>> &edge : edges) {
        cin >> edge.ss.ff >> edge.ss.ss >> edge.ff;
    }
    sort(edges.begin(), edges.end());
    for (pair<int, pair<int, int>> edge : edges) {
        int w = edge.ff, u = edge.ss.ff, v = edge.ss.ss;
        if (dsu.merge(u, v)) {
            ans += w;
            ++cnt;
        }
    }
    if (cnt == n - 1) {
        cout << ans << endl;
    } else {
        cout << "impossible" << endl;
    }
    return 0;
}

File: Kruskal算法求最小生成树.py
class DSU: # 并查集
    def __init__(self, n):
        self.pre = list(range(n + 1))
        self.siz = [1 for _ in range(n + 1)]
    
    def find(self, x):
        x, xcopy = self.pre[x], x
        while x != self.pre[x]:
            x = self.pre[x]
        while xcopy != x:
            self.pre[xcopy], xcopy = x, self.pre[xcopy]
        return x
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.siz[x] += self.siz[y]
        self.pre[y] = x
        return True
    
    def size(self, x):
        return self.siz[self.find(x)]
    
n, m = map(int, input().split())
ans, cnt = 0, 0
dsu = DSU(n + 1)
edges = []
for _ in range(m):
    edges.append(list(map(int, input().split())))
edges.sort(key = lambda x: x[2])
for u, v, w in edges:
    if dsu.merge(u, v):
        ans += w
        cnt += 1

if cnt == n - 1:
    print(ans)
else:
    print("impossible")
